### PR TITLE
Array: correct number of outputs in `apply_gufunc`

### LINKED ISSUE #7668: `apply_gufunc` bugs with meta inference
**What happened**:
`apply_gufunc` returns a tuple with the wrong number of outputs for multi-output signatures.

**What you expected to happen**:
`apply_gufunc` would return the same number of items as the signature implied.

**Minimal Complete Verifiable Example**:

```python
import dask.array as da
from dask.array import apply_gufunc
import numpy

>>> len(apply_gufunc(lambda a, b: (a.max(axis=1), b.max()), '(i,I),(I)->(i),()', da.ones((10, 10), chunks=-1), da.ones(10, chunks=-1), output_dtypes=['float64', 'float64']))
1
```
This should return a tuple of 2 items (one array and one scalar), as the signature implies.

The issue is that `meta_from_array` encounters a "zero-size array to reduction operation" error. This error used to propagate (which would have taken us down a [more useful codepath](https://github.com/dask/dask/blob/ac1bd05cfd40207d68f6eb8603178d7ac0ded922/dask/array/gufunc.py#L432-L448) in `apply_gufunc`), but as of recently (#6736) it just selects the first element from `args_meta`: https://github.com/dask/dask/blob/ac1bd05cfd40207d68f6eb8603178d7ac0ded922/dask/array/utils.py#L158-L161

@pentschev I wonder if this `meta = args_meta[0]` should be conditioned on `len(args_meta) = 1`? When there are multiple arguments, I think the only sensible behavior is to return None, since we don't know how the arguments will be combined. Even when there's only a single argument, I personally thing it's incorrect to assume in generality that the output type will be the same as the input type; an arbitrary user-defined function could do all sorts of other things before/after calling `np.min`.

Even if `meta_from_array` had raised an error, this still wouldn't have worked. This logic for generating meta from the `output_dtypes` https://github.com/dask/dask/blob/ac1bd05cfd40207d68f6eb8603178d7ac0ded922/dask/array/gufunc.py#L436-L440 will rarely/never be reached, because `output_dtypes` is typically a list, not a tuple.

Overall, there is inconsistency throughout `apply_gufunc` as to whether `nout`, the type+length of `output_dtypes`, or the type+length of `metas` is the source of truth for how many items to return.

**Anything else we need to know?**:

**Environment**:

- Dask version: ac1bd05cfd40207d68f6eb8603178d7ac0ded922
- Python version: 3.8.8
- Operating System: macOS
- Install method (conda, pip, source): source

### PR DESCRIPTION
**Pull request by gjoseph92**
Refactor `apply_gufunc` to compute meta itself, rather than relying on `blockwise` to do it. Having just one code path and consolidating everything on `meta` makes things easier to reason about, and happens to fix #7668. Additionally before, `func` would be called even if `output_dtypes` was given (contrary to the docstring); now it is no longer.

This also makes it an explicit error to pass both `output_dtypes` and `meta`—what you want in that case is rather ambiguous.

Also, use a modern function definition instead of popping things out of `kwargs` to support py2.

- [x] Closes #7668
- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

### CODE REVIEW

File: dask/array/tests/test_gufunc.py [hunk 1]
```diff
@@ -146,7 +146,7 @@ def stats(x):
 @pytest.mark.parametrize("vectorize", [False, True])
 def test_apply_gufunc_output_dtypes_string_many_outputs(vectorize):
     def stats(x):
-        return np.mean(x, axis=-1), np.std(x, axis=-1)
+        return np.mean(x, axis=-1), np.min(x, axis=-1)
```
Comment (gjoseph92) [inline dask/array/tests/test_gufunc.py:149]: This acts as a regression test against the example case in #7668 (`min`/`max` is required to trigger that codepath).

File: dask/array/tests/test_gufunc.py [hunk 2]
```diff
@@ -161,8 +161,8 @@ def foo(x, bar):
         assert bar == 2
         return x
 
-    ret = apply_gufunc(foo, "()->()", 1.0, output_dtypes="f", bar=2)
-    assert_eq(ret, np.array(1.0, dtype="f"))
+    ret = apply_gufunc(foo, "()->()", 1.0, output_dtypes=float, bar=2)
+    assert_eq(ret, np.array(1.0, dtype=float))
```
Comment (gjoseph92) [inline dask/array/tests/test_gufunc.py:165]: This was failing because the array chunks were float64, while meta expected float32 (`"f"`). I'm not sure why it ever passed before?

File: dask/array/utils.py [hunk 3]
```diff
@@ -65,13 +65,6 @@ def meta_from_array(x, ndim=None, dtype=None):
...
         x = x(shape=(0,) * (ndim or 0), dtype=dtype)
 
-    if (
-        not hasattr(x, "shape")
-        or not hasattr(x, "dtype")
-        or not isinstance(x.shape, tuple)
-    ):
-        return x
-
     if isinstance(x, list) or isinstance(x, tuple):
```
Comment (gjoseph92) [inline dask/array/utils.py:68]: I moved this before the `not hasattr(x, "shape") or not hasattr(x, "dtype")`, because otherwise it's a no-op: lists and tuples don't have `shape`/`dtype`, so we always would have returned from the function before reaching this check.

File: dask/array/utils.py [hunk 4]
```diff
@@ -156,8 +156,10 @@ def compute_meta(func, _dtype, *args, **kwargs):
                 else:
                     return None
             except ValueError as e:
-                # min/max functions have no identity, attempt to use the first meta
-                if "zero-size array to reduction operation" in str(e):
+                # min/max functions have no identity, just use the same input type when there's only one
+                if len(
+                    args_meta
+                ) == 1 and "zero-size array to reduction operation" in str(e):
```
Comment (gjoseph92) [inline dask/array/utils.py:162]: As noted in https://github.com/dask/dask/issues/7668. @pentschev, curious about your thoughts here. Personally, I don't think we should have this case at all; I don't think it's a safe assumption to make in generality for any user-defined function. But I do see the convenience.

Comment (pentschev) [reply dask/array/utils.py:162]: I agree with your assessment, t's not a safe assumption for user-defined functions indeed and I would as well prefer not to need this. However, without this condition attempting to compute meta for a 0-D downstream array (e.g., CuPy) will fall into the general `except Exception` case and `return None`, which will cause meta to be a NumPy array, and thus break those downstream use cases. The problem is we have no known alternative, this is why we check the specific exception message, for any other corner cases we would have to extend this function to handle them properly, but user-defined functions should guarantee proper handling of 0-D arrays to avoid such issues.

**Review by pentschev — COMMENTED** [review]
Thanks for pinging me here @gjoseph92 . Overall this looks good to me, it seems like a more feature-complete manner of handling `apply_gufunc`. I've left a few comments, and assuming I haven't missed anything about that, I think `test_apply_gufunc_pass_additional_kwargs ` is a potentially concerning test, so it may be worth checking whether some sneaky dtype casting is happening silently, such cases can happen and can be difficult to find, but perhaps it's just me being overly cautious.

I've also verified that this PR doesn't break the CuPy tests we have (which aren't covered by CI yet), so this feels like a good solution. Thanks for working on it!

File: dask/array/tests/test_gufunc.py (hunk 2 above)
Comment (gjoseph92) [reply dask/array/tests/test_gufunc.py:166]: @pentschev I dug into this and found a number of concerning things:
* On `main`, despite saying `output_dtypes="f"` (float32), `ret.dtype` is float64! As I mentioned in the description, `output_types` is currently actually ignored unless inferring the dtype from `func` fails. Since calling the `foo` function with a float64 succeeds (the `1.0` input becomes NumPy float64), `apply_gufunc` returns a float64 array, even though you told it to do float32.
* **_Then_, `assert_eq` doesn't actually check dtypes?!** Turns out there is actually nothing in the `assert_eq` logic to check whether the dtypes of the two computed arrays match—just lots of logic to check that each one's meta matches its result. So the fact that `ret` and the expected `np.array(1.0, dtype="f")` have different dtypes is ignored, and the test passes. @jrbourbeau is this well-known? Should I open an issue about it? Seems like misleading behavior.

Since my changes now respect `output_dtypes`, the resulting array's `meta` (float32) does not match its computed dtype (float64, from the `1.0` input value), and `assert_eq` complained about that meta mismatch until I made the change now in this PR.

So basically, the old test was wrong: the output dtype in this case is in fact float64, so passing `output_dtype="f"` should have resulted in an error from `assert_eq` about a meta mismatch, except that `output_dtype` was ignored.

File: dask/array/tests/test_gufunc.py (hunk 1 above)
Comment (gjoseph92) [reply dask/array/tests/test_gufunc.py]: I added `std` back.

File: dask/array/tests/test_gufunc.py (hunk 2 above)
Comment (pentschev) [reply dask/array/tests/test_gufunc.py:166]: > * On `main`, despite saying `output_dtypes="f"` (float32), `ret.dtype` is float64! As I mentioned in the description, `output_types` is currently actually ignored unless inferring the dtype from `func` fails. Since calling the `foo` function with a float64 succeeds (the `1.0` input becomes NumPy float64), `apply_gufunc` returns a float64 array, even though you told it to do float32.

To be honest, I'm not certain what's the exact behavior we want/users expect here. But if `output_dtypes` is specified, shouldn't `apply_gufunc` cast to the `output_dtypes`? Note that we do that in other places, such as [`partial_reduce`](https://github.com/dask/dask/blob/f547f081572cef0d2f99868d930d9256d2d9a00f/dask/array/reductions.py#L339). That would be, of course, only for the case where `meta` is computed automatically, but the user wants to enforce the `output_dtypes` nevertheless.

> * **_Then_, `assert_eq` doesn't actually check dtypes?!** Turns out there is actually nothing in the `assert_eq` logic to check whether the dtypes of the two computed arrays match—just lots of logic to check that each one's meta matches its result. So the fact that `ret` and the expected `np.array(1.0, dtype="f")` have different dtypes is ignored, and the test passes. @jrbourbeau is this well-known? Should I open an issue about it? Seems like misleading behavior.

Interesting catch, and I would say we should indeed open an issue. Even though I don't recall if there's a reason why we don't do it, it's worth at least discussing that, in case that is in fact a bug.

Comment (jsignell) [reply dask/array/tests/test_gufunc.py:166]: I aggree with your assessment that `assert_eq` doesn't check computed dtype against meta dtype. That'd be a nice thing to add :)

**Comment by jsignell** [comment]
Is this ready to be merged? Also I am planning on merging https://github.com/dask/dask/pull/6863 today as well - I think the two PRs address different issues.

**Comment by pentschev** [comment]
I'm fine with that @jsignell , I think the only unanswered portion here was the first part of https://github.com/dask/dask/pull/7669#discussion_r635063205 . It's only a matter of how we want/should treat `output_dtypes` when `meta` is computed automatically.

**Comment by gjoseph92** [comment]
I think the question is whether `output_dtypes` should cause a cast. I'm leaning towards it should, because I think it's more intuitive. Additionally, it's more similar to `np.vectorize`, where giving `otypes` does ensure the function outputs that type.

It's basically a question of whether `output_dtypes` is purely meant as a hint/annotation, to help dask determine the `meta`—and in general, we treat meta as something that's supposed to accurately simulate computation, but which does not affect computation in any way—or whether it's actually a request on the user's part.

However, I have very little preference, this is working as is, and based on the docstring for `output_dtypes`, I don't think the current behavior is wrong per se. So @jsignell, I'm happy if you want to merge this now. We could always add the cast as a separate PR in the future.

**Comment by mathause** [comment]
Unless I am missing something you currently don't have a test for `apply_gufunc` passing both `meta` and `output_dtypes`, i.e. the newly raised `ValueError` is not tested.

**Comment by gjoseph92** [comment]
@mathause good catch, there's no test for raising that error.
